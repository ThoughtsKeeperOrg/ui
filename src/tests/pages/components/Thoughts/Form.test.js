import {render, screen, container, fireEvent, act} from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import '@testing-library/jest-dom'


import CrudDataRepository from '../../../../repositories/crudDataRepository';

import ThoughtsForm from '../../../../pages/components/Thoughts/Form'


// https://testing-library.com/docs/react-testing-library/cheatsheet

// <ThoughtsForm addThought={addThought} dataRepository={dataRepository} />



// const items = [{id: 1, content: 'some text'}, {id: 2, content: 'some other text'} ]


const text = 'Some text';
const addThought = (item) => {};
// const stubRepository = () => ({
//     createItem: async (data) => {
//         // return params.content here

//       console.log(data)

//         return { 'entity': { id: 1, content: 'some content'}};
//     }
// });



// const fn = jest.fn().mockName('Unicorn')

// const dataRepository = stubRepository();

// const stubRepository = () => ({});

jest.mock('../../../../repositories/crudDataRepository');
// const dataRepository = jest.mocked(stubRepository);
const mockCreate = jest.fn();
mockCreate.mockImplementation();
CrudDataRepository.mockImplementation(() => {
  return {
    createItem: mockCreate,
  };
});
const dataRepository = CrudDataRepository();

function timeout(delay) {
    return new Promise( res => setTimeout(res, delay) );
}


test('renders form', async () => {
  const {container} = render(<ThoughtsForm addThought={addThought} dataRepository={dataRepository}/>)

  expect(container).toMatchSnapshot();
});

test('calls data repository to submit new entity', async () => {
  const {container} = render(<ThoughtsForm addThought={addThought} dataRepository={dataRepository}/>)
  const input = container.querySelector('textarea');

  act(() => {
    fireEvent.change(input, {target: {value: text}})
  });
  
  await act( async () => {
    fireEvent.click(screen.getByText('Submit'))
  });

  expect(container).toMatchSnapshot();
  expect(mockCreate).toBeCalledWith({"thought": {"content": "Some text"}})
});

test('sends file as base64 string', async () => {
  const {container} = render(<ThoughtsForm addThought={addThought} dataRepository={dataRepository}/>)

  const blob = new Blob(['str']);
  const file = new File([blob], 'values.json', { type: 'application/JSON'});
  File.prototype.text = jest.fn().mockResolvedValueOnce('str');

  const input = container.querySelector("[type='file']");

  await act( async () => {
    await userEvent.upload(input, file);
    await timeout(10);
  });


  await act( async () => {
    await userEvent.click(screen.getByText('Submit'));
    await timeout(10);
  });

  expect(mockCreate).toBeCalledWith({
      thought: { content: '' },
      file: {
        convert_to_text: true,
        filename: 'values.json',
        type: 'application/json',
        file_base64: 'c3Ry'
      }
  });
  
  expect(container).toMatchSnapshot();
});
