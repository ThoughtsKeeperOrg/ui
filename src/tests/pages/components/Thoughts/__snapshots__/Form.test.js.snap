// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calls data repository to submit new entity 1`] = `
<div>
  <form
    class=""
  >
    <div
      class="mb-3 input-group"
    >
      <textarea
        class="form-control"
        rows="2"
      />
    </div>
    <div
      class="mb-3 input-group"
    >
      <input
        class="form-control form-control-sm"
        type="file"
      />
      <select
        class="file_joined_select form-select form-select-sm"
      >
        <option
          value="true"
        >
          Convert to text
        </option>
        <option
          value="false"
        >
          Do not convert to text
        </option>
      </select>
    </div>
    <button
      class="btn btn-outline-primary"
      type="button"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`renders form 1`] = `
<div>
  <form
    class=""
  >
    <div
      class="mb-3 input-group"
    >
      <textarea
        class="form-control"
        rows="2"
      />
    </div>
    <div
      class="mb-3 input-group"
    >
      <input
        class="form-control form-control-sm"
        type="file"
      />
      <select
        class="file_joined_select form-select form-select-sm"
      >
        <option
          value="true"
        >
          Convert to text
        </option>
        <option
          value="false"
        >
          Do not convert to text
        </option>
      </select>
    </div>
    <button
      class="btn btn-outline-primary"
      type="button"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`sends file as base64 string 1`] = `
<div>
  <form
    class=""
  >
    <div
      class="mb-3 input-group"
    >
      <textarea
        class="form-control"
        rows="2"
      />
    </div>
    <div
      class="mb-3 input-group"
    >
      <input
        class="form-control form-control-sm"
        type="file"
      />
      <select
        class="file_joined_select form-select form-select-sm"
      >
        <option
          value="true"
        >
          Convert to text
        </option>
        <option
          value="false"
        >
          Do not convert to text
        </option>
      </select>
    </div>
    <button
      class="btn btn-outline-primary"
      type="button"
    >
      Submit
    </button>
  </form>
</div>
`;
